### PR TITLE
Use the latest JSON serialisation rules in Refract

### PR DESCRIPTION
**Pull request by kylef**
This pull request updates the JSON Serialiser to follow Refract 1.0 JSON Serialisation rules. This PR also includes commits that provide the old serialiser and old serialisation rules for API Elements. As you can see in the changes there was a lot of custom serialisation based on the element (`sourceMap`, `enum`) and even rules based on attribute key names (`default`) that differ from other attributes. 

- d2c74a1 - Updates to follow new serialisation rules
- 066aad1a939faec8833426054d26bb1f1ad6ce9f - Moves the old serialiser and old serialiser tests to the json-0.6 file. This commit does not change any of the existing code so the contents (other than file creation) do not need to be reviewed. Changes are done separately in follow up commits.
- https://github.com/refractproject/minim/pull/127/commits/cc0995a885946962af7eee14502fae412253ce61 - This commit changes the serialiser to be consistent with the attribute/meta serialisation rules that Drafter used.
- https://github.com/refractproject/minim/pull/127/commits/69db54dbd704e8358249cf4d78881bbf6acd64f7 is minor refactoring allowing the serialiser to contain a method to override the content serialiser for a given element type. This is used in the subsequent commits.
- https://github.com/refractproject/minim/pull/127/commits/13a29ad5ac446ce2f60021741df641fccd8d5dbc - Encode ref elements like Refract 0.6.
- e305796 - Wrap dataStructure content in arrays like Drafter v3
- https://github.com/refractproject/minim/pull/127/commits/1c7ec1827156f23f0bd580ec907d0318c22b5bba - Serialise metadata attributes as meta.

### Tasks

- [x] Test in Swagger - https://github.com/apiaryio/swagger-zoo/pull/39 updates the Swagger fixtures to use the new serialiser.
- [x] Test re-serialisation from new/old serialiser using Drafter fixtures.

Using the following script, we can take all of the test fixtures from Drafter and then re-serialise the elements using JSON 0.6 serialiser and compare to fixtures found in Drafter 3.x. This allows us to be sure that re-serialisation works exactly as the old serialised responses from Drafter.

```js
var fs = require('fs');
var glob = require('glob');
var differ = require('json-diff'); // https://github.com/andreyvit/json-diff
var minim = require('./lib/minim').namespace();
var JSON06Serialiser = require('./lib/serialisers/json-0.6');

var serialiser = new JSON06Serialiser();

var drafter3 = '/Users/kyle/Projects/apiary/drafter3'; // v3.2.7 source
var drafter4 = '/Users/kyle/Projects/apiary/drafter'; // master source

var ignored = [
  // https://github.com/apiaryio/drafter/pull/484
  '/test/fixtures/schema/array-restricted-to-type.json',
  '/test/fixtures/schema/array-restricted-to-types-complex.json',
  '/test/fixtures/schema/array-restricted-to-types.json',

  // https://github.com/apiaryio/drafter/pull/483
  '/test/fixtures/api/schema-custom-json.json',

  // drafter 3.x was incorrectly adding empty api category during error case
  '/test/fixtures/parse-result/error.json',
];

var fixtures = glob
  .sync(drafter4 + '/test/**/*.json')
  .map(path => path.replace(drafter4, ''))
  .filter(path => !ignored.includes(path));

var success = 0;
var failed = 0;
var skipped = 0;

for (var fixture of fixtures) {
  try {
    var oldFixture = require(drafter3 + fixture);
    var newFixture = require(drafter4 + fixture);
  } catch (error) {
    // Fixture is doesn't exist in old version (probably was ast named .ast.json)
    skipped++;
    continue;
  }

  if (oldFixture['_version']) {
    // Fixture is AST
    skipped++;
    continue;
  }

  // Deserialise the new fixture
  const element = minim.fromRefract(newFixture);

  // Re-serialise element as Refract 0.6
  const reserailised = serialiser.serialise(element);

  // Compare re-serialised against old fixture
  const oldFixtureJSON = JSON.stringify(oldFixture, null, 2);
  const reserialisedJSON = JSON.stringify(reserailised, null, 2)

  if (oldFixtureJSON !== reserialisedJSON) {
    const diff = differ.diffString(reserailised, oldFixture);
    process.stdout.write(diff);
    failed++;
  } else {
    success++;
  }
}

console.log(`${success} success, ${failed} failed and ${skipped} skipped.`);
```

### CODE REVIEW

**Comment by kylef** [comment]
Added "do not merge" label as I opened discussion on Slack regarding current Source Map structure which we should finalise before merging.

**Comment by pksunkara** [comment]
Please remove this commit from this PR and I will merge it. https://github.com/refractproject/minim/pull/127/commits/3d94c528db6db9f3e28c2828f3b2036ad617076d

**Comment by kylef** [comment]
@pksunkara removing that commit would mean we cannot re-serialise to the same output as Drafter 3 and thus cannot roll this out to the parsing services. Why would we remove this commit?

**Review by pksunkara — APPROVED** [review]

**Pull request merged**
